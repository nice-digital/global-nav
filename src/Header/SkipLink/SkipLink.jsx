import React from "react";
import PropTypes from "prop-types";
import styles from "./SkipLink.module.scss";

const SkipLink = (props) => {
	const handleClick = (e) => {
		const href = e.currentTarget.getAttribute("href");

		if (href && href.indexOf("#") === 0) {
			const id = href.split("#")[1],
				element = document.getElementById(id);

			if (element) {
				e.preventDefault();

				element.setAttribute("tabIndex", "-1");
				element.focus();
				element.scrollIntoView();
			}
		}
	};

	return (
		<a href={props.to} className={styles.link} onClick={handleClick}>
			{props.children}
		</a>
	);
};

//TODO Convert proptypes to typescript
SkipLink.propTypes = {
	to: PropTypes.string.isRequired,
	children: PropTypes.oneOfType([
		PropTypes.arrayOf(PropTypes.node),
		PropTypes.node,
	]).isRequired,
};

export default SkipLink;
