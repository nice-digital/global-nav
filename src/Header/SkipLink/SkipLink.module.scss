@use '@nice-digital/nds-core/scss/colours';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/utils';
@use '../header-settings';

%link {
  clip: rect(1px 1px 1px 1px);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;

  &:focus {
    background: #fff;
    clip: auto;
    color: colours.$link;
    display: block;
    height: auto;
    left: utils.rem(spacing.$x-small);
    outline: 3px solid colours.$focus;
    overflow: visible;
    padding: utils.rem(spacing.$small spacing.$medium);
    position: absolute;
    top: utils.rem(spacing.$x-small);
    white-space: nowrap;
    width: auto;
    z-index: header-settings.$base-z-index + 5;
  }
}

.link {
  @extend %link;
}
