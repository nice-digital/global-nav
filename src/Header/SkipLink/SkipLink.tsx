import React from "react";
import styles from "./SkipLink.module.scss";

interface SkipLinkProps {
	to: string;
	children: React.ReactNode;
}

const SkipLink = (props: SkipLinkProps) => {
	const handleClick = (e: React.MouseEvent) => {
		const href = e.currentTarget.getAttribute("href");

		if (href && href.indexOf("#") === 0) {
			const id = href.split("#")[1],
				element = document.getElementById(id);

			if (element) {
				e.preventDefault();

				element.setAttribute("tabIndex", "-1");
				element.focus();
				element.scrollIntoView();
			}
		}
	};

	return (
		<a href={props.to} className={styles.link} onClick={handleClick}>
			{props.children}
		</a>
	);
};

export default SkipLink;
