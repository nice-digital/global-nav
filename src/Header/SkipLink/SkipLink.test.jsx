import React from "react";
import SkipLink from "./SkipLink";
import { render } from "@testing-library/react";

describe("SkipLinks", () => {
	const skipLinkId = "test-id";
	let contentDiv;

	beforeEach(() => {
		// Create the skip target
		contentDiv = document.createElement("div");
		contentDiv.id = skipLinkId;
		document.body.append(contentDiv);

		// Fake window.scrollTo because it isn't implemented in jsdom AND we want to spy on it
		window.scrollTo = jest.fn();
	});

	afterEach(() => {
		Object.defineProperties(window.HTMLElement.prototype, {
			offsetTop: {
				get: () => 0,
			},
		});

		contentDiv.parentNode.removeChild(contentDiv);
	});

	it("Matches snapshot with a hash", () => {
		const { container } = render(<SkipLink to="#test">Skip</SkipLink>);
		expect(container).toMatchSnapshot();
	});

	it("Matches snapshot with an external destination", () => {
		const { container } = render(
			<SkipLink to="https://a11y.com">Skip</SkipLink>
		);
		expect(container).toMatchSnapshot();
	});

	it("Renders skip link id href from prop", () => {
		const { container } = render(<SkipLink to="#test">Skip</SkipLink>);
		expect(wrapper.find("a").at(0).props().href).toEqual("#test");
	});

	it("Renders an external link if supplied", () => {
		const { container } = render(
			<SkipLink to="https://a11y.com">Skip</SkipLink>
		);
		expect(wrapper.find("a").at(0).props().href).toEqual("https://a11y.com");
	});

	it("Triggers click handler on skip to content click", () => {
		const handleClick = jest.spyOn(SkipLink.prototype, "handleClick");
		const { container } = render(<SkipLink to="test">Skip</SkipLink>);

		wrapper
			.find("a")
			.at(0)
			.simulate("click", {
				currentTarget: {
					getAttribute: () => "#" + skipLinkId,
				},
				preventDefault: () => {},
			});

		expect(handleClick).toHaveBeenCalledTimes(1);
	});

	it("Prevents default and moves focus to the skip link target on click", () => {
		const { container } = render(<SkipLink to="test">Skip</SkipLink>);

		const preventDefault = jest.fn();

		wrapper.instance().getYOffset = jest.fn(() => 99);

		wrapper
			.find("a")
			.at(0)
			.simulate("click", {
				currentTarget: {
					getAttribute: () => "#" + skipLinkId,
				},
				preventDefault: preventDefault,
			});

		expect(preventDefault).toHaveBeenCalledTimes(1);
		expect(contentDiv.getAttribute("tabIndex")).toEqual("-1");
		expect(document.activeElement).toBe(contentDiv);
		expect(window.scrollTo).toHaveBeenCalledWith(0, 99);
	});

	describe("getYOffset", () => {
		it("should get offset top relative to window", () => {
			const { container } = render(<SkipLink to="test">Skip</SkipLink>);

			const element = {
				offsetLeft: 0,
				offsetTop: 99,
				scrollTop: 9,
				offsetParent: {
					offsetLeft: 0,
					offsetTop: 98,
					scrollTop: 8,
				},
			};

			expect(wrapper.instance().getYOffset(element)).toBe(180);
		});
	});
});
