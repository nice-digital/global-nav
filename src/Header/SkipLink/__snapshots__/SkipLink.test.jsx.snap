// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipLinks Matches snapshot with a hash 1`] = `
<div>
  <a
    class="link"
    href="#test"
  >
    Skip
  </a>
</div>
`;

exports[`SkipLinks Matches snapshot with an external destination 1`] = `
<div>
  <a
    class="link"
    href="https://a11y.com"
  >
    Skip
  </a>
</div>
`;
